src/on_chain/on_chain_surplus.py

```diff
@@ -0,0 +1,296 @@
+"""
+on_chain file
+"""
+from config import INFURA_KEY
+from contracts.gpv2_settlement import gpv2_settlement as gpv2Abi
+from typing import List, Tuple
+from fractions import Fraction
+from copy import deepcopy
+import requests
+import logging
+from web3 import Web3
+from src.on_chain.instance_file import instance1, instance2
+
+# GPv2 contract address
+address = "0x9008D19f58AAbD9eD0D60971565AA8510560ab41"
+
+
+class DecodedSettlement:
+    def __init__(
+        self,
+        tokens: List[str],
+        clearing_prices: List[int],
+        trades: List[Tuple[int, int, str, int, int, int, bytes, int, int, int, bytes]],
+        interactions: List[List[Tuple[str, int, bytes]]],
+    ):
+        self.tokens = tokens
+        self.clearing_prices = clearing_prices
+        self.trades = trades
+        self.interactions = interactions
+
+    @classmethod
+    def new(cls, contract_instance, transaction):
+        # Decode the function input
+        decoded_input = contract_instance.decode_function_input(transaction)[1:]
+        # Convert the decoded input to the expected types
+        tokens = decoded_input[0]["tokens"]
+        clearing_prices = decoded_input[0]["clearingPrices"]
+        trades = decoded_input[0]["trades"]
+        interactions = decoded_input[0]["interactions"]
+
+        # Create and return a new instance of DecodedSettlement
+        return cls(tokens, clearing_prices, trades, interactions)
+
+
+class OnChainEBBO:
+    def __init__(self):
+        """
+        Makes ethereum node connection to Infura, and creates contract instance
+        """
+        self.web_3 = Web3(
+            Web3.HTTPProvider(f"https://mainnet.infura.io/v3/{INFURA_KEY}")
+        )
+        self.contract_instance = self.web_3.eth.contract(address=address, abi=gpv2Abi)
+        self.logger = get_logger()
+
+    def get_hashes_by_block(self, start_block, end_block):
+        """
+        Function filters hashes by contract address
+        """
+        filter_criteria = {
+            "fromBlock": int(start_block),
+            "toBlock": int(end_block),
+            "address": address,
+        }
+        transactions = self.web_3.eth.filter(filter_criteria).get_all_entries()
+        settlement_hashes_list = []
+        for tx in transactions:
+            tx_hash = (tx["transactionHash"]).hex()
+            if tx_hash not in settlement_hashes_list:
```

Comment:
This was done because web3.py even fetches the transactions multiple event logs, and creates a separate entry for each event log.
This way ensures to avoid it from being added again (alternatively could use a set too). As another alt, will also try to see if something could be added to the filter.